pkg/webhook/pod/validating/pod_unavailable_budget.go

```diff
@@ -0,0 +1,146 @@
+)
+
+// parameters:
+// 1. allowed(bool) whether to allow this request
+// 2. reason(string)
+// 3. err(error)
+func (p *PodCreateHandler) podUnavailableBudgetValidatingPod(ctx context.Context, req admission.Request) (bool, string, error) {
+	var newPod, oldPod *corev1.Pod
+	var dryRun bool
+
+	newPod = &corev1.Pod{}
+	switch req.AdmissionRequest.Operation {
+	// filter out invalid Update operation, we only validate update Pod.MetaData, Pod.Spec
+	case admissionv1beta1.Update:
+		//decode new pod
+		err := p.Decoder.Decode(req, newPod)
+		if err != nil {
+			return false, "", err
+		}
+		oldPod = &corev1.Pod{}
+		if err = p.Decoder.Decode(
+			admission.Request{AdmissionRequest: admissionv1beta1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}},
+			oldPod); err != nil {
+			return false, "", err
+		}
+
+		options := &metav1.UpdateOptions{}
+		//decode eviction
+		err = p.Decoder.DecodeRaw(req.Options, options)
+		if err != nil {
+			return false, "", err
+		}
+		// if dry run
+		dryRun = dryrun.IsDryRun(options.DryRun)
+
+	// filter out invalid Delete operation, only validate delete pods resources
+	case admissionv1beta1.Delete:
+		if req.AdmissionRequest.SubResource != "" {
+			klog.V(6).Infof("pod(%s.%s) AdmissionRequest operation(DELETE) subResource(%s), then admit", req.Namespace, req.Name, req.SubResource)
+			return true, "", nil
+		}
+		deletion := &metav1.DeleteOptions{}
+		//decode eviction
+		err := p.Decoder.DecodeRaw(req.Options, deletion)
+		if err != nil {
+			return false, "", err
+		}
+		// if dry run
+		dryRun = dryrun.IsDryRun(deletion.DryRun)
+		key := types.NamespacedName{
+			Namespace: req.AdmissionRequest.Namespace,
+			Name:      req.AdmissionRequest.Name,
+		}
+		if err = p.Client.Get(ctx, key, newPod); err != nil {
+			return false, "", err
+		}
+
+	// filter out invalid Create operation, only validate create pod eviction subresource
+	case admissionv1beta1.Create:
+		// ignore create operation other than subresource eviction
+		if req.AdmissionRequest.SubResource != "eviction" {
+			klog.V(6).Infof("pod(%s.%s) AdmissionRequest operation(CREATE) Resource(%s) subResource(%s), then admit", req.Namespace, req.Name, req.Resource, req.SubResource)
+			return true, "", nil
+		}
+		eviction := &policy.Eviction{}
+		//decode eviction
+		err := p.Decoder.Decode(req, eviction)
+		if err != nil {
+			return false, "", err
+		}
+		// if dry run
+		dryRun = dryrun.IsDryRun(eviction.DeleteOptions.DryRun)
+		key := types.NamespacedName{
+			Namespace: req.AdmissionRequest.Namespace,
+			Name:      req.AdmissionRequest.Name,
+		}
+		if err = p.Client.Get(ctx, key, newPod); err != nil {
+			return false, "", err
+		}
+	}
+
+	isUpdated := req.AdmissionRequest.Operation == admissionv1beta1.Update
+	// returns true for pod conditions that allow the operation for pod without checking PUB.
+	if newPod.Status.Phase == corev1.PodSucceeded || newPod.Status.Phase == corev1.PodFailed ||
```

Comment:
for newly created pod, pod status maybe empty

Reply:
status is not a pointer, so there will not panic. when status is not initialized, newPod.Status.Phase equals "", the logic is also normal